crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -42,19 +49,32 @@ pub use bevy_ecs_macros::States;
 /// }
 ///
 /// ```
-pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug {}
+pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug {
+    /// How many other states this state depends on.
+    /// Used to help order transitions and de-duplicate [`ComputedStates`], as well as prevent cyclical
+    /// `ComputedState` dependencies.
+    const DEPENDENCY_DEPTH: usize = 1;
+}
 
-/// The label of a [`Schedule`](super::Schedule) that runs whenever [`State<S>`]
+/// This trait allows a state to be mutated directly using the [`NextState<S>`] resource.
+///
+/// This is in contrast with [`ComputedStates`], which do not allow modification - they are
+/// automatically derived.
+///
+/// It is implemented as part of the [`States`] derive, but can also be added manually.
+pub trait FreelyMutableState: States {}
```

Comment:
Unfortunately, nothing.

Because of the lack of specialization, we can't auto-implement this for States that aren't computed while also preventing it from being implemented from those that are and allowing it for SubStates. As a result - it has to be implemented in the macro, which unfortunately puts it in user-land.

Are you aware of any ways to restrict this kind of use?

Reply:
Sealed traits won't help us here unfortunately: because we're using macros, there's no way to limit visibility to *only* auto-generated code.